addons/website_sale_loyalty/tests/test_ewallet.py

```diff
@@ -0,0 +1,70 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import http
+from odoo.tests.common import HttpCase
+from odoo.tests import tagged
+from odoo.addons.website.tools import MockRequest
+from odoo.addons.website_sale_loyalty.controllers.main import WebsiteSale
+
+@tagged('post_install', '-at_install')
+class TestEwallet(HttpCase):
+
+    def setUp(self):
+        super().setUp()
+
+        self.WebsiteSaleController = WebsiteSale()
+        self.website = self.env['website'].browse(1)
+
+        self.carpet = self.env['product.product'].create({
+            'name': 'Test Carpet',
+            'list_price': 500.0,
+            'website_published': True,
+        })
+
+        self.topup = self.env['product.product'].create({
+            'name': 'Ewallet Top up',
+            'list_price': 50.0,
+            'website_published': True,
+        })
+
+        self.empty_order = self.env['sale.order'].create({
+            'partner_id': self.env.user.partner_id.id
+        })
+
+        self.ewallet_program = self.env['loyalty.program'].create([{
+            'name': 'E-wallet Card Program',
+            'program_type': 'ewallet',
+            'trigger': 'auto',
+            'applies_on': 'future',
+            'rule_ids': [(0, 0, {
+                'reward_point_mode': 'money',
+                'reward_point_amount': 10,
+                'product_ids': self.topup,
+            })],
+            'reward_ids': [(0, 0, {
+                    'discount_mode': 'per_point',
+                    'discount': 1,
+                    'discount_applicability': 'order',
+                })],
+        }])
+        installed_modules = set(self.env['ir.module.module'].search([
+            ('state', '=', 'installed'),
+        ]).mapped('name'))
+        for _ in http._generate_routing_rules(installed_modules, nodb_only=False):
+            pass
```

Comment:
is this necessary? I remember I had to do it once for a specific test, but it was a 'hacky' solution and it should only be done if you need it.

Reply:
If I dont't iclude this, I get "if routing['type'] == 'http':  # _generate_routing_rules() ensures type is set
KeyError: 'type'"